Add ADC sequence sample

wkhadgar opened the pull request:
The current ADC sample using "channel" DTS construction lacks flexibility in the sequence use. For example, if I create a `struct adc_dt_spec` like the sample says:
```c
/* Data of ADC io-channels specified in devicetree. */
static const struct adc_dt_spec adc_channels[] = {
	DT_FOREACH_PROP_ELEM(DT_PATH(zephyr_user), io_channels,
			     DT_SPEC_AND_COMMA)
};
```
It is clear that each `struct adc_dt_spec` represents a channel from my DTS. There is no problem with that.

The issue happens when you try to make full use of the `struct adc_sequence` capabilities. Since you can specify more than one channel to read, let's say:
```c
struct adc_sequence sequence = {
	.channels = BIT(0) | BIT(1) | BIT(2),
};
```
Passing the created channel as an argument of `adc_read_dt(&adc_channels[i], &sequence)` with a sequence that requires more than one channel (and possibly not the one you passed on the API call) is unintuitive and error-leading. What's the sense of passing a channel to read other channels?
To solve this, it feels natural to pass not the channel, but the ADC device as the `dev` argument of `adc_read` (and not the `_dt` version of it), and then make full use of `struct adc_sequence`.

This PR fits this purpose, showing a neat, clear way, to get the full value of `struct adc_sequence` on a batch reading, without losing coherence on the API calls.

The previous sample was moved to a new `adc_dt` directory.

wkhadgar committed:
samples: drivers: adc: add adc sequence sample

Add sequence usage sample for adc driver.

Signed-off-by: Paulo Santos <pauloroberto.santos@edge.ufal.br>

wkhadgar commented:
Worth noting that even though this PR may look oversized, 82 of the files were just moved to a subdirectory in order from having another ADC sample. Revision and code changes are located in the second commit, where I actually make the new sample.

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,110 @@
+#define ADC_NODE DT_ALIAS(adc)
+
+/* Get the ADC resolution from one of its children. */
+#define ADC_RESOLUTION DT_PROP(DT_CHILD(ADC_NODE, channel_0), zephyr_resolution)
```
dont like this requirement to name a channel a certain way, even some of the overlays added in this PR are not compliant with this and are not building (like mimxrt1040_evk), just read the resolution from the DT cfg in the main function

decsny replied:
I think hardcoded is not good because each ADC may not support that resolution, instead make a kconfig and use .conf files for the supported boards of the sample to make sure that a resolution can be used that is supported for each adc

wkhadgar replied:
Sure thing. Updated to `.conf`.

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,110 @@
+#define ADC_RESOLUTION DT_PROP(DT_CHILD(ADC_NODE, channel_0), zephyr_resolution)
+
+/* Number of samples to be made on the sequence. */
+#define SEQUENCE_SAMPLES 5
```
kconfig

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,110 @@
+	DT_FOREACH_CHILD(ADC_NODE, ADC_CHANNEL_CONFIG_DT_AND_COMMA)};
+
+/* Options for the sequence sampling. */
+const struct adc_sequence_options options = {
```
static I guess

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,110 @@
+		.buffer_size = sizeof(channel_reading),
+		.channels = BIT(channel_cfgs[CHANNEL_A].channel_id) |
+			    BIT(channel_cfgs[CHANNEL_B].channel_id),
+		.resolution = ADC_RESOLUTION,
```
read from channel_cfgs struct

decsny replied:
okay, I put this comment because originally it was being read from DT in the top of the file

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,110 @@
+		.channels = BIT(channel_cfgs[CHANNEL_A].channel_id) |
+			    BIT(channel_cfgs[CHANNEL_B].channel_id),
+		.resolution = ADC_RESOLUTION,
+		.options = &options,
```
why is options a static struct but the sequence is on the stack?

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,110 @@
+	};
+
+	if (!device_is_ready(adc)) {
+		printk("ADC controller device %s not ready\n", adc->name);
```
use printf instead of printk throughout sample

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,110 @@
+	CHANNEL_A = 0,
+	CHANNEL_B,
+
+	CHANNEL_COUNT,
```
why hardcode the number of channels?

decsny replied:
what I mean is, can't you count the number of channels from DT, and set the sequence.channels using a for loop

wkhadgar replied:
Gotcha, I was thinking on a sample with fixed number of channels, throughout the boards, but this seems really better. Updated the initialization to account for that.

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,110 @@
+		err = adc_read(adc, &sequence);
+		if (err < 0) {
+			printk("Could not read (%d)\n", err);
+			continue;
```
maybe instead of going to the next iteration right away it should also do the sleep in this case?

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,110 @@
+
+			printk("- %s, channel %"PRId32", %"PRId32" sequence samples:\n",
+			       adc->name, channel_cfgs[i].channel_id, SEQUENCE_SAMPLES);
+			for (size_t j = 0U; j < SEQUENCE_SAMPLES; j++) {
```
use more descriptive loop variables than i and j

wkhadgar committed:
samples: drivers: adc: naming adc sample

The adc sample was in the root directory of the adc samples.
Move all the files to a subdirectory adc_dt.

Signed-off-by: Paulo Santos <pauloroberto.santos@edge.ufal.br>

decsny commented on samples/drivers/adc/adc_sequence/src/main.c:
```diff
@@ -0,0 +1,105 @@
+static const struct device *adc = DEVICE_DT_GET(ADC_NODE);
+
+/* Data array of ADC channels for the specified ADC. */
+const struct adc_channel_cfg channel_cfgs[] = {
```
static

wkhadgar replied:
Fixed.

kartben commented on samples/drivers/adc/adc_sequence/README.rst:
```diff
@@ -0,0 +1,66 @@
+notice that the for the whole sequence reading, only one call for the ``adc_read`` API is made.
+If voltage of the used reference can be obtained, the raw readings are converted to millivolts.
+
+This example constructs an adc device and setup its channels, according to the
```
```suggestion
This example constructs an adc device and setups its channels, according to the
```

kartben reviewed (dismissed):
+1 for sample docs, thanks for implementing the suggested changes!

wkhadgar reviewed (commented):
@decsny @kartben 
Just rebased the branch with the upstream and solved the conflicts (some overlays were conflicting)
![image](https://github.com/zephyrproject-rtos/zephyr/assets/51432914/303aba5d-357f-4cfb-a760-781a6722d26f)

rjmaris commented:
I have tried to get it working on `nucleo_f429zi` board. As long as I specify only one channel for the sequence (in the overlay), the results are as expected. With default 5 samples per channel.
But when I select more than one channel, the `adc_read()` is getting blocked - also with e.g. 1 or 2 samples per channel. No error à la -22 or whatever else is reported:
```
*** Booting Zephyr OS build v3.6.0-3796-g6b048fdf7109 ***
ADC sequence reading [0]:      # comment: failing multi-channel approach
*** Booting Zephyr OS build v3.6.0-3796-g6b048fdf7109 ***
ADC sequence reading [0]:
- adc@40012000, channel 3, 2 sequence samples:
- - 588 = 0mV    # comment: the failing mV is just a matter of missing mv
- - 663 = 0mV
ADC sequence reading [1]:
- adc@40012000, channel 3, 2 sequence samples:
- - 400 = 0mV
- - 500 = 0mV
```

For the purpose of this experiment I added an overlay file for the board - see patch file. The additions/changes are based upon recent commit 6b048fd.

When I inspect the code execution using breakpoints, the function `set_sequencer()` in `drivers/adc/adc_stm32.c` is iterated correctly, including correct processing of the set channel bits. After the setup is finished, I see no return from `adc_context_wait_for_completion(&data->ctx)`. This is quite curious because there is no problem when only one channel is sequenced.
[0001-samples-drivers-adc_sequence-for-nucleo_f429-board.txt](https://github.com/zephyrproject-rtos/zephyr/files/15299344/0001-samples-drivers-adc_sequence-for-nucleo_f429-board.txt) (_no patch-suffix:_ [_If you use Linux and try to upload a .patch file, you will receive an error message. This is a known issue._](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files))

rjmaris commented:
@wkhadgar - today, a **nucleo C031C6** board arrived here. I encounter the same problem as with my added `nucleo_f429zi` overlay. Even in commit `6173f71` (samples: drivers: adc: naming adc sample), see below extract from latest commit and the commit just mentioned.

```
*** Booting Zephyr OS build v3.6.0-3797-gf04f5183f415 ***
ADC sequence reading [0]:
*** Booting Zephyr OS build v3.6.0-2899-g6173f71d42cd ***
ADC sequence reading [0]:
```

wkhadgar commented:
@rjmaris I took a note on that, made some static code review, and it seems to be something with the STM driver implementation...

Will try to debug the cause for this behavior, since the Zephyr side is configured accordingly to the `sequence` use, it falls on the API implementation to process and act as expected, reading the data as the `sequence` has been set to.

Since the call stack up to `adc_context_wait_for_completion(&data->ctx)` is ok (thus no error), I suspect the problem to be ISR related.

rjmaris commented:
@wkhadgar  Thanks for your quick response! First of all: this sequencer approach is indeed the only reasonable way to process several channels, instead of the former approach. But where you recall the `adc_context_wait...` this now makes me aware of the fact that the efficiency gain (and approachable sample rate) related to the former approach is within certain limits, because the caller's program flow is halted during this wait.

Am I right in my guess that not all boards with overlay in this sequencer approach were actually tested to be functional? The reason that I ordered a **nucleoC031C6** after persisting failing operation with my project board (STM32F4xx) is simply because it's a STM32 board.

At the moment I hardly can imagine to get the problem solved because of too little Zephyr experience. Perhaps a contributor to *zephyrproject* from ST could be contacted, and asked to look at this.

rjmaris commented:
Meanwhile I had found a FRDM board in my office (the move to another office space gave rise to questions where stuff has been located). It's a FRDM-KL25z, but it may be expected that the sequencer will operate on all Kinetis based boards.

Because the order of sampling in sequence mode is not intuitively visible in the output, I have swapped both `for` loops. And made detail changes. It looks like this now:
```
*** Booting Zephyr OS build v3.6.0-3454-g0a4f1236bb06 ***
ADC sequence reading [0], 5 sequence samples:
      adc0.8       adc0.9       adc0.12      adc0.13
 1: 1025=0001mV  4095=0003mV  0002=0000mV  0260=0000mV  
 2: 0789=0000mV  4095=0003mV  0002=0000mV  0267=0000mV  
 3: 0763=0000mV  4095=0003mV  0001=0000mV  0315=0000mV  
 4: 0934=0000mV  4095=0003mV  0001=0000mV  0410=0000mV  
 5: 0697=0000mV  4095=0003mV  0001=0000mV  0277=0000mV  
ADC sequence reading [1], 5 sequence samples:
      adc0.8       adc0.9       adc0.12      adc0.13
 1: 0265=0000mV  4095=0003mV  0002=0000mV  0237=0000mV  
 2: 0648=0000mV  4095=0003mV  0001=0000mV  0459=0000mV  
 3: 0807=0000mV  4095=0003mV  0001=0000mV  0406=0000mV  
 4: 0841=0000mV  4095=0003mV  0001=0000mV  0490=0000mV  
 5: 0865=0000mV  4095=0003mV  0001=0000mV  0429=0000mV  
ADC sequence reading [2], 5 sequence samples:
      adc0.8       adc0.9       adc0.12      adc0.13
 1: 0267=0000mV  4095=0003mV  0002=0000mV  0243=0000mV  
 2: 0407=0000mV  4095=0003mV  0001=0000mV  0243=0000mV  
 3: 0565=0000mV  4095=0003mV  0001=0000mV  0245=0000mV  
 4: 0783=0000mV  4095=0003mV  0000=0000mV  0472=0000mV  
 5: 0830=0000mV  4095=0003mV  0001=0000mV  0551=0000mV  
```
See attached patch.
**Note:** the '0' in 'adc0' should be made automatically right when channels might be attached to another adc unit. I don't know how to accomplish that (I remember dat in earlier Zephyr versions (pre 3.3?) this was dons so in the adc sample.
[0001-samples-drivers-adc-adc_sequence-represent-output-ta.patch.txt](https://github.com/user-attachments/files/15938021/0001-samples-drivers-adc-adc_sequence-represent-output-ta.patch.txt)

wkhadgar commented:
@rjmaris that is awesome! Loved the grouped view of the channels, really makes clearer how the sampling is made!
